### DISCUSSION TITLE
mobc connection pool

### DISCUSSION THREAD

**Discussion by byblakeorriver** [weiznich/diesel_async#181] [Q&A] [answered]
Hello.
I am trying to track down why I am receiving these errors in my services only after many hours have passed: 
```
Input/output error: Input/output error: Driver error: `Connection to the server is closed.
```

Here is my setup, and I am running in rust version 1.78.0
```
diesel = { version = "2.2.2", features = ["chrono"] }
diesel-async = { version = "0.5.0", features=["mysql", "mobc"] }
mobc = "0.8.4"
```

Here is how I am using it:
```
let mut manager_config = ManagerConfig::default();
manager_config.recycling_method = RecyclingMethod::Fast;
let manager = AsyncDieselConnectionManager::<AsyncMysqlConnection>::new_with_config(
  DATABASE_URL.clone(),
  manager_config,
);
let pool = Pool::builder()
  .max_open(*DB_POOL_MAX_SIZE)
  .max_idle(*DB_POOL_MAX_IDLE)
  .max_idle_lifetime(Some(Duration::from_secs(
    *DB_POOL_MAX_IDLE_LIFETIME_SECONDS,
  )))
  .test_on_check_out(*DB_POOL_TEST_ON_CHECKOUT)
  .health_check_interval(Some(Duration::from_secs(
    *DB_POOL_HEALTH_CHECK_INTERVAL_SECONDS,
  )))
  .build(manager);
let pool = Arc::new(pool);
```

I then store that Arc in a context that gets shared throughout my service.

I guess my questions are:
1. Has anyone else seen this behavior?
2. Do you know if this a mobc thing, or a diesel-async while using mobc thing?
3. If it is a diesel-async thing, are there any tools that I can use to help debug this so that I can take a look.

If you need more info please let me know.

**Comment by weiznich** [comment, marked as answer]
I have not seen this behavior yet, so I'm not sure where it's coming from. I cannot really say what's the best way to debug this.

That written: You configure your pool to use `RecyclingMethod::Fast`, which essentially skips a check whether the connection is still usable. You might want to prefer using `RecyclingMethod::Verified` (the default), which performs a check query before assuming that the connection still works. In your particular case it might be that the connection is just closed by the server or by any network interruption. `RecyclingMethod::Verified` would detect this case and open a new connection instead.

**Reply by byblakeorriver** [reply]
Good call!

I will try that and report back.

Thank you!

**Reply by byblakeorriver** [reply]
I have switched to using `RecyclingMethod::Verified`

After a few hours I was seeing the same error so I don't think it is that.
I am going to continue to try and dig. I will let you know what I find.

**Reply by byblakeorriver** [reply]
I think it has to do with another bit of my configuration. I had test_on_checkout set to false.
I think that was preventing the checks from happening even with the recycling_method set to Verified.
Going to test my hypothesis and will update once I know for sure.

**Comment by byblakeorriver** [comment]
Ok, I did finally get around this issue and it was entirely a configuration issue on my end.
Once that was fixed, I did end up getting the same errors in this issue: https://github.com/weiznich/diesel_async/issues/139

I am going to use the work around you mentioned for now, and try to get a PR in to use the ping he suggested.
Thank you for your help!
